libs/filer/files_client.go

```diff
@@ -165,40 +212,138 @@ func (w *FilesClient) Delete(ctx context.Context, name string, mode ...DeleteMod
 		return CannotDeleteRootError{}
 	}
 
-	err = w.apiClient.Do(ctx, http.MethodDelete, urlPath, nil, nil, nil)
-
-	// Return early on success.
-	if err == nil {
-		return nil
+	// Issue a stat call to determine if the path is a file or directory.
+	info, err := w.Stat(ctx, name)
+	if err != nil {
+		return err
 	}
 
-	// Special handling of this error only if it is an API error.
-	var aerr *apierr.APIError
-	if !errors.As(err, &aerr) {
+	// Issue the delete call for a directory
+	if info.IsDir() {
+		err = w.workspaceClient.Files.DeleteDirectoryByDirectoryPath(ctx, absPath)
+
+		// The directory delete API returns a 400 if the directory is not empty
+		var aerr *apierr.APIError
+		if errors.As(err, &aerr) && aerr.StatusCode == http.StatusBadRequest {
+			return DirectoryNotEmptyError{absPath}
+		}
 		return err
 	}
 
-	// This API returns a 404 if the specified path does not exist.
-	if aerr.StatusCode == http.StatusNotFound {
+	// Issue the delete call for a file
+	err = w.workspaceClient.Files.DeleteByFilePath(ctx, absPath)
+
+	// This files delete API returns a 404 if the specified path does not exist.
+	var aerr *apierr.APIError
+	if errors.As(err, &aerr) && aerr.StatusCode == http.StatusNotFound {
 		return FileDoesNotExistError{absPath}
 	}
+	return err
+}
 
-	// This API returns 409 if the underlying path is a directory.
-	if aerr.StatusCode == http.StatusConflict {
-		return DirectoryNotEmptyError{absPath}
+func (w *FilesClient) recursiveDelete(ctx context.Context, name string) error {
+	filerFS := NewFS(ctx, w)
+	dirsToDelete := make([]string, 0)
+	callback := func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+
+		// Files API does not allowing deleting non-empty directories. We instead
+		// collect the directories to delete and delete them once all the files have
+		// been deleted.
+		if d.IsDir() {
+			dirsToDelete = append(dirsToDelete, path)
+			return nil
+		}
+
+		return w.delete(ctx, path)
 	}
 
-	return err
+	// Walk the directory and delete all the files.
+	err := fs.WalkDir(filerFS, name, callback)
+	if err != nil {
+		return err
+	}
+
+	// Delete the directories in reverse order to ensure that the parent
+	// directories are deleted after the children. This is possible because
+	// fs.WalkDir walks the directories in lexicographical order.
+	for i := len(dirsToDelete) - 1; i >= 0; i-- {
+		err := w.delete(ctx, dirsToDelete[i])
```

Comment:
Done! We now delete files in parallel. Harder to do for directories because we need to go from inner to outer when deleting directories. Skipping for now since the number of directors will likely be an order of magnitude less than the number of files.